### PR TITLE
Tuscan square detectors

### PR DESCRIPTION
**Pull request by mmk-1**
Hello, this branch includes the additional tuscan-intra-class & tuscan-inter-class detection types plus the tests for tuscan squares (using junit theory), tuscan-only-class, tuscan-intra-class & tuscan-inter-class.

### CODE REVIEW

**Comment by winglam** [comment]
@mmk-1 Can you please resolve the merge conflicts?

**Comment by mmk-1** [comment]
Hello, I have resolved the conflicts.

File: idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/TestShuffler.java [hunk 1]
```diff
@@ -234,4 +233,52 @@ public List<String> alphabeticalAndTuscanOrder(int count, boolean isTuscan) {
...
+
+    public List<String> tuscanIntraClassOrder(int round) {
+        List<String> classes = new ArrayList<>(classToMethods.keySet());
+        HashMap<String, int[][]> classToPermutations = new HashMap<String, int[][]>();
+        Collections.sort(classes);
+        final List<String> fullTestOrder = new ArrayList<>();
+        int n = classes.size(); // n is number of classes
+        int[][] classOrdering = Tuscan.generateTuscanPermutations(n);
+        for (String className : classes) {
+            // This condition may be needed for the tests
```
Comment (august782) [inline idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/TestShuffler.java:245]: What is the meaning behind the commented out code here?

File: idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/TestShuffler.java [hunk 2]
```diff
@@ -234,4 +233,52 @@ public List<String> alphabeticalAndTuscanOrder(int count, boolean isTuscan) {
...
+            //     classToPermutations.put(className, methodPermuation);
+            // } else {
+            int[][] methodPermuation = Tuscan.generateTuscanPermutations(classToMethods.get(className).size());
+            classToPermutations.put(className, methodPermuation);
+            // }
+        }
+        HashMap<String, List<String>> newClassToMethods = new HashMap<String, List<String>>();
+        List<String> permClasses = new ArrayList<String>();
+        int classRound = round;
+        while ((classOrdering.length - 1) < classRound) {
```
Comment (august782) [inline idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/TestShuffler.java:259]: Should add some comment to explain the reasoning/intention behind this loop concerning the class round, so we remember later what is the goal.

File: idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/DetectorFactory.java [hunk 3]
```diff
@@ -28,6 +29,8 @@ public static Detector makeDetector(final InstrumentingSmartRunner runner, final
             return new TuscanOnlyClassDetector(runner, baseDir, rounds, detectorType(), tests);
         } else if (detectorType().startsWith("alphabetical")) {
             return new AlphabeticalDetector(runner, baseDir, rounds, detectorType(), tests);
+        } else if (detectorType().equals("tuscan-intra-class")) {
```
Comment (august782) [inline idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/DetectorFactory.java:32]: Maybe move this above to put next to tuscan (or move tuscan below to be next to this one).

File: idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/TuscanIntraClassDetector.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            }
+        }
+        return classes.size();
+    }
+    
+    public static int findMaxMethodSize(HashMap<String, List<String>> classToMethods) {
+        int maxMethodSize = 0;
+        for (String className : classToMethods.keySet()) {
+            int nn = classToMethods.get(className).size();
+            if(maxMethodSize < nn) {
```
Comment (august782) [inline idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/TuscanIntraClassDetector.java:39]: Space after `if`

File: idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/TuscanIntraClassDetector.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        classToMethods = new HashMap<>();
+        for (final String test : tests) {
+            final String className = TestShuffler.className(test);
+            if (!classToMethods.containsKey(className)) {
+                classToMethods.put(className, new ArrayList<>());
+            }
+            classToMethods.get(className).add(test);
+        }
+        int classSize = classToMethods.keySet().size();
+        // if (classSize == 3 || classSize == 5) {
```
Comment (august782) [inline idflakies-core/src/main/java/edu/illinois/cs/dt/tools/detection/detectors/TuscanIntraClassDetector.java:57]: Is it unnecessary now to consider the 3 and 5 edge cases?

File: idflakies-core/src/test/java/edu/illinois/TuscanIntraClassTest.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+import org.junit.Assert;
+
+import edu.illinois.cs.dt.tools.detection.TestShuffler;
+import edu.illinois.cs.dt.tools.detection.detectors.TuscanIntraClassDetector;
+
+public class TuscanIntraClassTest {
+    @Test
+    public void test() throws Exception {
+        String[] testArray = {
+        "cn.edu.hfut.dmic.webcollector.util.CharsetDetectorTest.testGuessEncodingByMozilla",
```
Comment (august782) [inline idflakies-core/src/test/java/edu/illinois/TuscanIntraClassTest.java:19]: I think these strings should be indented one more.

File: idflakies-core/src/test/java/edu/illinois/TuscanIntraClassTest.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+                        visitedMethodPairs.add(newPair);
+                    }
+                }
+                currentMethodOrdering.clear();
+            }
+        }
+        Assert.assertEquals(allMethodPairs.size(), visitedMethodPairs.size());
+    }    
+
+    public static HashMap<String, List<String>> generateClassToMethods (List<String> tests) {
```
Comment (august782) [inline idflakies-core/src/test/java/edu/illinois/TuscanIntraClassTest.java:95]: No space after method name.
